### PR TITLE
CRM-20274 Added credit card type field for contribution repeat report

### PR DESCRIPTION
**Pull request by Edzelopez**
* [CRM-20274: Add credit card type field for Contribution Repeat Report](https://issues.civicrm.org/jira/browse/CRM-20274)

### CODE REVIEW

File: CRM/Report/Form/Contribute/Repeat.php [hunk 1]
```diff
@@ -400,6 +418,16 @@ public function from() {
        .{$this->contributionJoinTableColumn}
 LEFT JOIN $this->tempTableRepeat2 {$this->_aliases['civicrm_contribution']}2
        ON {$this->groupByTableAlias}.$fromCol = {$this->_aliases['civicrm_contribution']}2.{$this->contributionJoinTableColumn}";
+
+    // for credit card type
+    if ($this->isTableSelected('civicrm_financial_trxn')) {
```
Comment (eileenmcnaughton) [inline CRM/Report/Form/Contribute/Repeat.php]: This section is repeated so many times I'm inclined to stick it in a function $this->joinFinancialTrxnFromContribution or something.

**Comment by monishdeb** [comment]
Tested the patch works for me. 

@eileenmcnaughton in terms of optimization, I think we should raise a separate issue because its not only about ```civicrm_financial_trxn``` but there are other build-from-clause section which are repeated across different reports, for other table, like ```civicrm_email```, ```civicrm_phone``` etc. 
I have prepared a partial patch https://gist.github.com/monishdeb/71ca5436c037d9a3407bf04e97756ef4 of how we can achieve that, but in my opinion, we should resolve it in separate ticket as lot of refactoring changes involved across different reports. What ya say ?

**Comment by eileenmcnaughton** [comment]
@monishdeb my worry here is that we have a bunch of patches adding the same code to multiple reports, so it would be better to get it right for the added code. Re your patch - I'm not at all keen on it being based on a big switch statement. Switch statements are often discussed as a code smell and while there can be all sorts of debate on that I think that switch statement will get messy really quickly. I think just having a function to encapsulate the oft-repeated 5 lines of code (like we do with address) is going to work out cleaner.

I don't have a problem with you reviewing & merging these report changes we are comfortable with what the criteria is. I think it is low-risk & the JMA-reviews-JMA is offset by the fact you are fresh to this code. 

From a risk pov the main concern I have is to ensure each report is already & always doing a group by - because otherwise the group_concat does. That is just a per-report check.

Stylistically, in a addition to trying not to repeat the lines of code on the join I also think that based on the comments in CRM_Core_PseudoConstant::get function we should .. do what it says & call buildOptions

**Comment by monishdeb** [comment]
I have discussed a DB error for this report with @Edzelopez and we agreed on excluding CC type field and filter from this Repeat Report. So closing this PR now.

**Pull request closed without merging**